include/Oscillator.h

```diff
@@ -153,6 +164,63 @@ class LMMS_EXPORT Oscillator
 		return m_userWave->userWaveSample( _sample );
 	}
 
+	struct wtSampleControl {
+		float frame;
+		f_cnt_t f1;
+		f_cnt_t f2;
+		int band;
+	};
+
+	inline wtSampleControl getWtSampleControl(const float sample)
+	{
+		wtSampleControl control;
+		control.frame = sample * OscillatorConstants::WAVETABLE_LENGTH;
+		control.f1 = static_cast<f_cnt_t>(control.frame) % OscillatorConstants::WAVETABLE_LENGTH;
+		if (control.f1 < 0)
+		{
+			control.f1 += OscillatorConstants::WAVETABLE_LENGTH;
+		}
+		control.f2 = control.f1 < OscillatorConstants::WAVETABLE_LENGTH - 1 ?
+					control.f1 + 1 :
+					0;
+		control.band = waveTableBandFromFreq(m_freq * m_detuning * Engine::mixer()->processingSampleRate());
+		return control;
+	}
+
+	inline sample_t wtSample(const sample_t table[OscillatorConstants::WAVE_TABLES_PER_WAVEFORM_COUNT][OscillatorConstants::WAVETABLE_LENGTH], const float _sample)
+	{
+		wtSampleControl control = getWtSampleControl(_sample);
+		return linearInterpolate(table[control.band][control.f1],
+				table[control.band][control.f2], fraction(control.frame));
+	}
+
+	inline sample_t wtSample(const std::unique_ptr<OscillatorConstants::waveform_t>& table, const float sample)
+	{
+		wtSampleControl control = getWtSampleControl(sample);
+		return linearInterpolate((*table)[control.band][control.f1],
+				(*table)[control.band][control.f2], fraction(control.frame));
+	}
+
+	inline sample_t wtSample( sample_t **table, const float _sample)
+	{
+		wtSampleControl control = getWtSampleControl(_sample);
+		return linearInterpolate(table[control.band][control.f1],
+				table[control.band][control.f2], fraction(control.frame));
+	}
+
+	inline int  waveTableBandFromFreq(float freq)
+	{
+		int band = (69 + static_cast<int>(ceil(12.0f * log2f(freq / 440.0f)))) / OscillatorConstants::SEMITONES_PER_TABLE;
```

Comment:
I think it is probably as you describe: better lose one of the harmonics near the limit rather than risk aliasing. Although in this case it would not really hurt if `round()` was used instead: the limit in `OscillatorConstants.h` is set to 20 kHz, so even the lowest sample rate available for export (44.1 kHz) would easily handle harmonics slightly over 20 kHz without aliasing.

Maybe I could also add that to a comment.

Reply:
Agreeing to all.